docs/cosmology/io.rst

```diff
@@ -65,12 +62,154 @@ method. The method uses this cosmology class, regardless of the class
+``to_format`` / ``from_format``.
+
+.. EXAMPLE START: Planck18 to mapping and back
+
+>>> from astropy.cosmology import Planck18
+>>> cm = Planck18.to_format("mapping")
+>>> cm
+{'cosmology': <class 'astropy.cosmology.core.FlatLambdaCDM'>,
+ 'name': 'Planck18',
+ 'H0': <Quantity 67.66 km / (Mpc s)>,
+ 'Om0': 0.30966,
+ ...
+
+Now this dict can be used to load a new cosmological instance identical
+to the ``Planck18`` cosmology from which it was generated.
+
+>>> from astropy.cosmology import Cosmology
+>>> cosmo = Cosmology.from_format(cm, format="mapping")
+>>> cosmo == Planck18
+True
+
+.. EXAMPLE END
+
+
+
+.. _custom_cosmology_converters:
+
+Custom Cosmology To/From Formats
+================================
+
+Custom representation formats may also be registered into the Astropy Cosmology
+I/O framework for use by these methods. For details of the framework see
+:ref:`io_registry`.
+
+.. EXAMPLE START : custom to/from format
+
+As an example, the following is an implementation of an `astropy.table.Row`
+converter. Note that we can use other registered parsers -- here "mapping"
+-- to make the implementation much simpler.
+
+We start by defining the function to parse a `astropy.table.Row` into a
+`~astropy.cosmology.Cosmology`. This function should take 1 positional
+argument, the row object, and 2 keyword arguments, for how to handle
+extra metadata and which Cosmology class to use. Details about metadata
+treatment are in ``Cosmology.from_format.help("mapping")``.
+
+.. code-block:: python
+    :emphasize-lines: 14,15
+
+    >>> import copy
+    >>> from astropy.cosmology import Cosmology
+    
+    >>> def from_table_row(row, *, move_to_meta=False, cosmology=None):
+    ...     # get name from column
+    ...     name = row['name'] if 'name' in row.columns else None
+    ...     meta = copy.deepcopy(row.meta)
+    ...     # turn row into mapping (dict of the arguments)
+    ...     mapping = {}
+    ...     mapping["cosmology"] = meta.pop("cosmology")
+    ...     mapping.update({k: v for k, v in zip(row.colnames, row.values())})
+    ...     mapping["meta"] = meta
+    ...     # build cosmology from map
+    ...     return Cosmology.from_format(mapping,
+    ...                                  move_to_meta=move_to_meta, cosmology=cosmology)
+
+The next step is a function to perform the reverse operation: parse a
+`~astropy.cosmology.Cosmology` into a `~astropy.table.Row`. This function
+requires only the cosmology object and a ``*args`` to absorb unneeded
+information passed by `astropy.io.registry.UnifiedReadWrite` (which
+implements `astropy.cosmology.Cosmology.to_format`).
+
+.. code-block:: python
+    :emphasize-lines: 4
+
+    >>> from astropy.table import QTable
+    
+    >>> def to_table_row(cosmology, *args):
+    ...     p = cosmology.to_format("mapping")
+    ...     # create metadata from mapping
+    ...     meta = p.pop("meta")
+    ...     meta["cosmology"] = p.pop("cosmology").__name__
```

Comment:
This line is confusing. Is it a requirement that the Table / Row representation have `meta['cosmology'] = <cosmology_name>`? To put in another way, sloshing the `cosmology` into `meta` for this representation seems weird. Why can't it just be in the main params `p` as a string?

Reply:
Ah, yeah. It probably should be for this example. In a previous PR (now closed) I had implemented a ``table`` format where the Cosmology was stored in meta. The thinking there was that a ``table`` would allow an MCMC chain of cosmology parameters, where each iteration was a row. This is normally done with a numpy array, but that doesn't do units or metadata. As the cosmology never changes it's metadata, but the "name" column could be the index. That was too much for this example, so long story short I think you're right! Changing it now.

Reply:
Done!